### PR TITLE
Nosql - mongo backend

### PR DESCRIPTION
**Pull request by jonhealy1**
**Related Issue(s):** #316

**Description:**
This PR adds a mongodb backend to stac-fastapi. It can also serve as an example for people wanting to add other nosql custom backends to their own projects. #322 

Pagination has not been implemented yet. There are going to things in this PR where people will wonder why something isn't done another - probably better - way and that is the type of feedback I am looking for. I will maintain this code and keep it up to date if it's approved. 

**PR Checklist:**

- [x] Code is formatted and linted (run `pre-commit run --all-files`)
- [x] Tests pass (run `make test`)
- [x] Documentation has been updated to reflect changes, if applicable, and docs build successfully (run `make docs`)
- [ ] Changes are added to the [CHANGELOG](https://github.com/stac-utils/pystac/blob/master/CHANGES.md).

### CODE REVIEW

File: stac_fastapi/mongo/setup.py [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+    desc = f.read()
+
+install_requires = [
+    "attrs",
+    "pydantic[dotenv]",
+    "stac_pydantic==2.0.*",
+    "stac-fastapi.types",
+    "stac-fastapi.api",
+    "stac-fastapi.extensions",
+    "shapely",
```
Comment (vincentsarago) [inline stac_fastapi/mongo/setup.py:15]: is shapely really used? (I couldn't find after a quick 👀)

File: stac_fastapi/mongo/stac_fastapi/mongo/core.py [hunk 2]
```diff
@@ -0,0 +1,354 @@
...
+        poly = [
+            [
+                [float(bbox[0]), float(bbox[1])],
+                [float(bbox[2]), float(bbox[1])],
+                [float(bbox[2]), float(bbox[3])],
+                [float(bbox[0]), float(bbox[3])],
+                [float(bbox[0]), float(bbox[1])],
+            ]
+        ]
+        return poly
```
Comment (vincentsarago) [inline stac_fastapi/mongo/stac_fastapi/mongo/core.py:153]: geojson_pydantic (one of stac-fastapi requirement) as a Polygon class that let you create a polygon from bounds
```python
from geojson_pydantic.geometries import Polygon
poly = Polygon.from_bounds(*bbox).dict(exclude_none=True)
```

File: stac_fastapi/mongo/setup.py (hunk 1 above)
Comment (jonhealy1) [reply stac_fastapi/mongo/setup.py:15]: It was used in a test - I switched to use geojson_pydantic like you suggested. Thank you.

File: docker-compose.mongo.yml [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+      - RELOAD=false
+      - ENVIRONMENT=local
+      - MONGO_USER=dev
+      - MONGO_PASS=stac
+      - MONGO_PORT=27017
+      - MONGO_HOST=172.17.0.1
+      - WEB_CONCURRENCY=10
+      - VSI_CACHE=TRUE
+      - GDAL_HTTP_MERGE_CONSECUTIVE_RANGES=YES
+      - GDAL_DISABLE_READDIR_ON_OPEN=EMPTY_DIR
```
[the remark this answers is not in the document]
Comment (jonhealy1) [reply docker-compose.mongo.yml:24]: Thanks. I erased these too. They seem to be only used for pgstac?
- DB_MIN_CONN_SIZE=1
- DB_MAX_CONN_SIZE=1

**Comment by vincentsarago** [comment]
@jonhealy1 I've removed myself from the reviewer. I'm not the best one to review this big piece of work 🙏

**Comment by jonhealy1** [comment]
> 

@vincentsarago Thanks for the feedback 🙏

**Pull request closed without merging**